File: CnCDemo/Tutorials/Java/Download2Go1_6/app/src/main/java/com/penthera/download2go1_6/AssetDetailActivity.java
/*
    Created by Penthera on 10/10/19.
    Copyright © 2019 penthera. All rights reserved.

    This source file contains a very basic example showing how to use the Penthera Download2Go SDK.
    Pay close attention to code comments marked IMPORTANT
*/
package com.penthera.download2go1_6;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.penthera.virtuososdk.client.IAssetManager;
import com.penthera.virtuososdk.client.IIdentifier;
import com.penthera.virtuososdk.client.ISegmentedAsset;
import com.penthera.virtuososdk.client.Observers;
import com.penthera.virtuososdk.client.Virtuoso;

import java.net.MalformedURLException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class AssetDetailActivity extends AppCompatActivity {

    public final static String ASSET_ID_KEY = "asset_id";

    private Virtuoso virtuoso;
    IAssetManager assetManager;

    private AssetQueueObserver assetQueueObserver;

    ISegmentedAsset asset;

    String assetId = null;

    private TextView titleView;
    private TextView uuidView;
    private TextView idView;
    private TextView fileTypeView;
    private TextView expectedSizeView;
    private TextView currentSizeView;
    private TextView durationView;
    private TextView statusView;
    private TextView pathView;
    private TextView playbackURLView;
    private TextView segmentCountView;
    private TextView firstPlayView;

    private static SimpleDateFormat dateFormatter = new SimpleDateFormat("MM/dd/yyyy HH:mm:ss a", Locale.US);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.detail_activity);

        titleView = findViewById(R.id.assetTitleValue);
        uuidView = findViewById(R.id.uuidValue);
        idView = findViewById(R.id.idValue);
        fileTypeView = findViewById(R.id.fileTypeValue);
        expectedSizeView = findViewById(R.id.expectedSizeValue);
        currentSizeView = findViewById(R.id.currentSizeValue);
        durationView = findViewById(R.id.durationValue);
        statusView = findViewById(R.id.statusValue);
        pathView = findViewById(R.id.pathValue);
        playbackURLView = findViewById(R.id.playbackValue);
        segmentCountView = findViewById(R.id.segmentCountValue);
        firstPlayView = findViewById(R.id.firstPlayValue);

        findViewById(R.id.btn_play).setOnClickListener(v -> playAsset());

        findViewById(R.id.btn_delete).setOnClickListener(v -> {
            if (asset != null) {
                assetManager.delete(asset);
            }
            finish();
        });

        // Initialise the SDK
        initVirtuosoSDK(savedInstanceState);

        if (savedInstanceState != null) {
            assetId = savedInstanceState.getString(ASSET_ID_KEY);
        }

    }

    @Override
    public void onResume() {
        super.onResume();

        // Resume the Download2Go SDK on activity resume
        virtuoso.onResume();
        virtuoso.addObserver(assetQueueObserver);


        if (assetId != null && assetManager != null) {
            List<IIdentifier> assets = assetManager.getByAssetId(assetId);
            if (assets != null && assets.size() > 0) {
                asset = (ISegmentedAsset) assets.get(0);
            }
        }
        updateUI();
    }

    @Override
    public void onPause() {
        super.onPause();

        // Pause the Download2Go SDK on activity pause
        virtuoso.onPause();
        virtuoso.removeObserver(assetQueueObserver);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(ASSET_ID_KEY, assetId);
    }

    public void initVirtuosoSDK(Bundle savedInstanceState) {
        // An alternative here is to create a singleton instance somewhere like the Application class
        // and share it between Activities.
        virtuoso = new Virtuoso(this);
        assetManager = virtuoso.getAssetManager();

        if (assetId == null) {
            Intent intent = getIntent();
            assetId = intent.getStringExtra(ASSET_ID_KEY);
        }

        List<IIdentifier> assets = assetManager.getByAssetId(assetId);
        if (assets != null && assets.size() > 0) {
            asset = (ISegmentedAsset) assets.get(0);
        }

        // set up observer
        assetQueueObserver = new AssetQueueObserver(this);
    }

    private void updateUI() {
        if (asset != null) {
            titleView.setText(asset.getMetadata());
            uuidView.setText(asset.getUuid());
            idView.setText(asset.getAssetId());
            fileTypeView.setText(fileTypeFromId(asset.segmentedFileType()));
            expectedSizeView.setText(String.format(Locale.getDefault(),"%.2f MB", asset.getExpectedSize()/1048576.00));
            currentSizeView.setText(String.format(Locale.getDefault(),"%.2f MB", asset.getCurrentSize()/1048576.00));
            durationView.setText(String.format(Locale.getDefault(),"%d seconds", asset.getDuration()));
            statusView.setText(MainActivity.getStatusText(this,asset.getDownloadStatus()));
            pathView.setText(asset.getLocalBaseDir());
            try {
                URL url = asset.getPlaybackURL();
                if (url != null) {
                    playbackURLView.setText(url.toString());
                } else {
                    playbackURLView.setText(R.string.unavailable);
                }
            } catch (MalformedURLException mue) {
                playbackURLView.setText(R.string.unavailable);
            }
            segmentCountView.setText(Integer.toString(asset.getTotalSegments()));
            long firstPlayTime = asset.getFirstPlayTime();
            if (firstPlayTime > 0) {
                firstPlayView.setText(dateFormatter.format(new Date(firstPlayTime*1000)));
            } else {
                firstPlayView.setText(R.string.not_yet_played);
            }
        }
    }

    private String fileTypeFromId(int fileTypeCode) {
        String fileTypeText = getString(R.string.unknown_type);
        switch (fileTypeCode) {
            case ISegmentedAsset.SEG_FILE_TYPE_HLS:
                fileTypeText = getString(R.string.hls_type);
                break;
            case ISegmentedAsset.SEG_FILE_TYPE_MPD:
                fileTypeText = getString(R.string.mpd_type);
                break;
        }
        return fileTypeText;
    }

    private void playAsset() {
        if (asset != null) {
            VideoPlayerActivity.playVideoDownload(this, asset);
        }
    }

    /**
     * This class observes the SDK download queue and provides updates during the download process.
     */
    static class AssetQueueObserver implements Observers.IQueueObserver {

        private AssetDetailActivity activity;

        public AssetQueueObserver(AssetDetailActivity activity) {
            this.activity = activity;
        }

        @Override
        public void engineStartedDownloadingAsset(IIdentifier asset) {
            updateItem(asset, true);
        }

        @Override
        public void enginePerformedProgressUpdateDuringDownload(IIdentifier asset) {
            updateItem(asset, false);
        }

        @Override
        public void engineCompletedDownloadingAsset(IIdentifier asset) {
            updateItem(asset, true);
        }

        @Override
        public void engineEncounteredErrorDownloadingAsset(IIdentifier asset) {
            // The base implementation does nothing.  See class documentation.
        }

        @Override
        public void engineUpdatedQueue() {
            // The base implementation does nothing.  See class documentation.
        }

        @Override
        public void engineEncounteredErrorParsingAsset(String s) {
            // The base implementation does nothing.  See class documentation.
        }

        private void updateItem(final IIdentifier identifier, boolean forceUpdate) {
            if (identifier instanceof ISegmentedAsset) {
                final ISegmentedAsset asset = (ISegmentedAsset) identifier;
                String assetId = asset.getAssetId();

                if (!TextUtils.isEmpty(assetId) && assetId.equals(activity.assetId)) {
                    activity.runOnUiThread(() -> {
                        activity.asset = asset;
                        activity.updateUI();
                    });
                }
            }
        }
    }
}

File: CnCDemo/Tutorials/Kotlin/Download2Go1_6/app/src/main/java/com/penthera/download2go1_6/AssetDetailActivity.kt
package com.penthera.download2go1_6

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.penthera.download2go1_6.databinding.DetailActivityBinding
import com.penthera.virtuososdk.client.IIdentifier
import com.penthera.virtuososdk.client.ISegmentedAsset
import com.penthera.virtuososdk.client.Observers.IQueueObserver
import com.penthera.virtuososdk.client.Virtuoso
import java.net.MalformedURLException
import java.text.SimpleDateFormat
import java.util.*

class AssetDetailActivity : AppCompatActivity() {
    private lateinit var virtuoso: Virtuoso

    private var assetQueueObserver: AssetQueueObserver? = null
    var asset: ISegmentedAsset? = null
    var assetId: String? = null

    private lateinit var binding: DetailActivityBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DetailActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.btnPlay.setOnClickListener { playAsset() }
        binding.btnDelete.setOnClickListener {
            if (asset != null) {
                virtuoso.assetManager.delete(asset)
            }
            finish()
        }

        // Initialise the SDK
        initVirtuosoSDK()
        if (savedInstanceState != null) {
            assetId = savedInstanceState.getString(ASSET_ID_KEY)
        }
    }

    public override fun onResume() {
        super.onResume()

        // Resume the Download2Go SDK on activity resume
        virtuoso.onResume()
        virtuoso.addObserver(assetQueueObserver)

        virtuoso.assetManager?.let{
            assetId?.let{ asset_id ->
                val assets = it.getByAssetId(asset_id)
                assets?.let{list ->
                    if(list.isNotEmpty()){
                        asset = list[0] as ISegmentedAsset
                    }
                }
            }
        }

        updateUI()
    }

    public override fun onPause() {
        super.onPause()

        // Pause the Download2Go SDK on activity pause
        virtuoso.onPause()
        virtuoso.removeObserver(assetQueueObserver)
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(ASSET_ID_KEY, assetId)
    }

    private fun initVirtuosoSDK() {
        // An alternative here is to create a singleton instance somewhere like the Application class
        // and share it between Activities.
        virtuoso = Virtuoso(this)
        if (assetId == null) {
            assetId = intent.getStringExtra(ASSET_ID_KEY)
        }
        val assets = virtuoso.assetManager?.getByAssetId(assetId)
        assets?.let{
            if (it.isNotEmpty()){
                asset = it[0] as ISegmentedAsset
            }
        }

        // set up observer
        assetQueueObserver = AssetQueueObserver(this)
    }

    @SuppressLint("SetTextI18n")
    private fun updateUI() {
        
        asset?.let{
            binding.assetTitleValue.text = it.metadata
            binding.uuidValue.text = it.uuid
            binding.idValue.text = it.assetId
            binding.fileTypeValue.text = fileTypeFromId(it.segmentedFileType())
            binding.expectedSizeValue.text = String.format("%.2f MB", it.expectedSize / 1048576.00)
            binding.currentSizeValue.text = String.format("%.2f MB", it.currentSize / 1048576.00)
            binding.durationValue.text = String.format("%d seconds", it.duration)
            binding.statusValue.text = MainActivity.getStatusText(this, it.downloadStatus)
            binding.pathValue.text = it.localBaseDir
            try {
                val url = it.playbackURL
                if (url != null) {
                    binding.playbackValue.text = url.toString()
                } else {
                    binding.playbackValue!!.setText(R.string.unavailable)
                }
            } catch (mue: MalformedURLException) {
                binding.playbackValue!!.setText(R.string.unavailable)
            }
            binding.segmentCountValue.text = it.totalSegments.toString()
            val firstPlayTime = it.firstPlayTime
            if (firstPlayTime > 0) {
                binding.firstPlayValue.text = dateFormatter.format(Date(firstPlayTime * 1000))
            } else {
                binding.firstPlayValue.setText(R.string.not_yet_played)
            }
        }
    }

    private fun fileTypeFromId(fileTypeCode: Int): String {
        var fileTypeText = getString(R.string.unknown_type)
        when (fileTypeCode) {
            ISegmentedAsset.SEG_FILE_TYPE_HLS -> fileTypeText = getString(R.string.hls_type)
            ISegmentedAsset.SEG_FILE_TYPE_MPD -> fileTypeText = getString(R.string.mpd_type)
        }
        return fileTypeText
    }

    private fun playAsset() {
        asset?.let{
            VideoPlayerActivity.playVideoDownload(it,this )
        }
    }

    /**
     * This class observes the SDK download queue and provides updates during the download process.
     */
    internal class AssetQueueObserver(private val activity: AssetDetailActivity) :
        IQueueObserver {
        override fun engineStartedDownloadingAsset(asset: IIdentifier) {
            updateItem(asset)
        }

        override fun enginePerformedProgressUpdateDuringDownload(asset: IIdentifier) {
            updateItem(asset)
        }

        override fun engineCompletedDownloadingAsset(asset: IIdentifier) {
            updateItem(asset)
        }

        override fun engineEncounteredErrorDownloadingAsset(asset: IIdentifier) {}

        override fun engineUpdatedQueue() {}

        override fun engineEncounteredErrorParsingAsset(s: String) {}

        private fun updateItem(identifier: IIdentifier?) {
            identifier?.let{
                if(it is ISegmentedAsset){
                    if(it.assetId.isNotEmpty() && it.assetId == activity.assetId){
                        activity.runOnUiThread {
                            activity.asset = it
                            activity.updateUI()
                        }
                    }
                }
            }
        }

    }

    companion object {
        const val ASSET_ID_KEY = "asset_id"
        @SuppressLint("SimpleDateFormat")
        private val dateFormatter = SimpleDateFormat("MM/dd/yyyy HH:mm:ss a")
    }
}